### PR TITLE
fix: Required properties warning now checked in child refs

### PR DESCRIPTION
**Pull request by rmkeezer**
Required property definitions are now checked for in objects that have allOfs, anyOfs, and refs. Also if they are not defined it is only a warning since "required" properties are not required to be defined anywhere.

I added an extra util function for reducing an object (resolving it's refs) since I did not see an easy way of getting an object in the jsSpec using only the ref path. Let me know if there is an easier way.

### CODE REVIEW

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 1]
```diff
@@ -39,11 +40,40 @@ module.exports.validate = function({ jsSpec }) {
       if (Array.isArray(obj.required)) {
         obj.required.forEach((requiredProp, i) => {
           if (!obj.properties || !obj.properties[requiredProp]) {
-            messages.addMessage(
-              path.concat([`required[${i}]`]).join('.'),
-              "Schema properties specified as 'required' must be defined",
-              'error'
-            );
+            // For allOf, ensure atleast one of the child objects has the required property
+            if (obj.allOf) {
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:44]: Might be a good idea to verify that this is an array. I know it is supposed to be according to the spec, but it would add some type safety since the code will crash on `forEach` otherwise.

File: test/plugins/validation/2and3/items-required-for-array-objects.test.js [hunk 2]
```diff
@@ -173,32 +173,46 @@ describe('validation plugin - semantic - items required for array objects - Open
...
+    expect(res.warnings.length).toEqual(1);
+    expect(res.warnings[0].path).toEqual(
       'components.schemas.TestObject.required[0]'
     );
-    expect(res.errors[0].message).toEqual(
-      "Schema properties specified as 'required' must be defined"
+    expect(res.warnings[0].message).toEqual(
+      "Schema properties specified as 'required' should be defined"
     );
-    expect(res.warnings.length).toEqual(0);
```
Comment (dpopp07) [inline test/plugins/validation/2and3/items-required-for-array-objects.test.js:201]: We need to have tests for all four scenarios: `allOf`, `oneOf`, `anyOf`, or none of the above.

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 3]
```diff
@@ -39,11 +40,40 @@ module.exports.validate = function({ jsSpec }) {
...
+              var props = {};
+              obj.allOf.forEach((childObj) => {
+                childObj = reduceObj(jsSpec, childObj);
+                props = Object.assign(props, childObj.properties);
+              });
+              if (!props[requiredProp]) {
+                messages.addMessage(
+                  path.concat([`required[${i}]`]).join('.'),
+                  "Schema properties specified as 'required' should be defined",
+                  'warning'
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:54]: I'm thinking we should make the a configurable check, instead of hard coding a "warning", and maybe setting the default value to error. What do you think?

File: src/plugins/utils/reduceObj.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    obj['allOf'].forEach((childObj) => {
+      newObj = Object.assign(newObj, reduceObj(jsSpec, childObj));
+    });
+    return newObj;
+  }
+  return obj;
+};
+
+module.exports = reduceObj;
+
```
Comment (dpopp07) [inline src/plugins/utils/reduceObj.js:17]: I believe there is an easier way. In other validators, we use `lodash/at` to resolve references. See [this line](https://github.com/IBM/openapi-validator/blob/d1ed5c0eef42f95aaac6befc80ae02a9fdb5bb04/src/plugins/validation/2and3/semantic-validators/walker-ibm.js#L36) for an example. You should be able to do this with the `jsSpec`. You may need some additional logic but you can just add a simple helper function in `items-required-for-array-objects.js` rather than creating a new utility file.

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js (hunk 3 above)
Comment (rmkeezer) [reply src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js]: It's a good idea to make it configurable, but i think setting the default value to "warning" would be better since the original problem statement, issue #1920, wanted to reduce importance of this issue from error to warning. @mkistler what do you think?

File: src/plugins/utils/reduceObj.js (hunk 4 above)
Comment (rmkeezer) [reply src/plugins/utils/reduceObj.js]: That's what I was looking for, great! Although it didn't actually reduce my logic much, I still have to convert the path from `"#/path/to/ref"` to `['path', 'to', 'ref']` and then reduce any $refs or allOfs

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js (hunk 3 above)
Comment (dpopp07) [reply src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js]: Oh good call - I missed that issue. Warning sounds good to me 👍

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 5]
```diff
@@ -36,14 +52,44 @@ module.exports.validate = function({ jsSpec }) {
...
+              }
+            // For anyOf/oneOf, ensure all of the child objects have the required property
+            } else if (Array.isArray(obj.anyOf) || Array.isArray(obj.oneOf)) {
+              var childList = obj.anyOf || obj.oneOf;
+              childList.forEach((childObj) => {
+                childObj = reduceObj(jsSpec, childObj);
+                if (childObj.properties && !childObj.properties[requiredProp]) {
+                  messages.addMessage(
+                    path.concat([`required[${i}]`]).join('.'),
+                    "Schema properties specified as 'required' should be defined",
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:81]: My concern here is that, if an API definition defines a `oneOf` with three schemas, and two of them are missing the definition, they will get two identical messages without distinguishing which ones are problematic.

Perhaps we should use a message like "Schema should define required property <property name>" and point the path at the schema itself

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 6]
```diff
@@ -10,8 +10,24 @@
 
 const { walk } = require('../../../utils');
 const MessageCarrier = require('../../../utils/messageCarrier');
+const at = require('lodash/at');
 
-module.exports.validate = function({ jsSpec }) {
+const reduceObj = function(jsSpec, obj) {
+  if (obj['$ref']) {
+    var objPath = obj['$ref'].split('/');
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:17]: I meant to comment on this earlier, but we shouldn't use `var` anywhere. These should all be `const`.

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 7]
```diff
@@ -10,8 +10,24 @@
...
+    objPath.shift();
+    return reduceObj(jsSpec, at(jsSpec, [objPath])[0]);
+  } else if (obj['allOf']) {
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:20]: What if the object is a `oneOf` or `anyOf`? I'm not sure what you would want to do here but I'm not sure it would work as intended if that happened.

File: test/plugins/validation/2and3/items-required-for-array-objects.test.js [hunk 8]
```diff
@@ -190,15 +191,141 @@ describe('validation plugin - semantic - items required for array objects - Open
...
+    expect(res.warnings.length).toEqual(1);
+    expect(res.warnings[0].path).toEqual(
+      'components.schemas.TestObject.required[0]'
+    );
+    expect(res.warnings[0].message).toEqual(
+      "Schema properties specified as 'required' should be defined"
+    );
+  });
+
+  it('should return an warning when a model with an anyOf does not define a required property in all of its children', () => {
```
Comment (dpopp07) [inline test/plugins/validation/2and3/items-required-for-array-objects.test.js:247]:
```suggestion
  it('should return a warning when a model with an anyOf does not define a required property in all of its children', () => {
```

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 9]
```diff
@@ -36,14 +52,44 @@ module.exports.validate = function({ jsSpec }) {
...
+      const checkRequiredProps = config.operations.check_required_properties;
       if (Array.isArray(obj.required)) {
         obj.required.forEach((requiredProp, i) => {
           if (!obj.properties || !obj.properties[requiredProp]) {
-            messages.addMessage(
-              path.concat([`required[${i}]`]).join('.'),
-              "Schema properties specified as 'required' must be defined",
-              'error'
-            );
+            // For allOf, ensure atleast one of the child objects has the required property
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:59]:
```suggestion
            // For allOf, ensure at least one of the child objects has the required property
```

File: README.md [hunk 10]
```diff
@@ -397,6 +397,7 @@ The default values for each rule are described below.
 | no_array_responses           | error   |
 | parameter_order              | warning |
 | operation_id_naming_convention | warning |
+| check_required_properties    | warning |
```
Comment (dpopp07) [inline README.md:400]: The naming convention for rules is usually to describe what the issue is. For example, `no_array_responses` leads to `error`.

I think we should name this something like `undefined_required_properties`

Comment (dpopp07) [reply README.md:400]: Also there's one more place in the README that needs to be filled in - the description for each rule

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js (hunk 7 above)
Comment (dpopp07) [reply src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:20]: I think it might also be good to have a test for the scenario of an `anyOf` here to make sure its covered.

Thinking about it now, I wonder if using the `resolvedSpec` would make all of this go away haha

Comment (rmkeezer) [reply src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:20]: I took a look at using resolvedSpec, but it seems to resolve just $refs and not allOf/oneOf/anyOf unless i'm using it wrong. In this case I dont think oneOfs or anyOfs can be feasibly "reduced" here since for allOf i can just combine all the properties of the children but for oneOfs and anyOfs, each child must be checked that it defines the required property.

Comment (dpopp07) [reply src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:20]: That's correct, it would resolve all $refs, including the refs inside of allOf/oneOf/anyOf, but no it wouldn't reconcile all of the sub-schemas. It still might be helpful.

I get that's the trouble with oneOf/anyOf, but I'm not sure the solution should be to ignore them. Maybe this function should be modified to return an array of schemas and in the case of single schemas or allOf, you just return one element. Then `forEach` can be called on the result. Just an idea. I guess this would only happen if sub-schemas to an allOf/oneOf/anyOf were _themselves_ a allOf/oneOf/anyOf, at which point the schema is just obnoxiously complicated... but I can't help but feel we should be complete here, even for the extreme scenario

File: test/plugins/validation/2and3/items-required-for-array-objects.test.js [hunk 11]
```diff
@@ -190,17 +191,293 @@ describe('validation plugin - semantic - items required for array objects - Open
...
+        }
+      }
+    };
+
+    const res = validate({ jsSpec: spec }, config);
+    expect(res.errors.length).toEqual(0);
+    expect(res.warnings.length).toEqual(0);
+  });
+
+  it('should not return a warning when a anyOf model with a nested allOf defines a required property in atleast one of its children', () => {
```
Comment (dpopp07) [inline test/plugins/validation/2and3/items-required-for-array-objects.test.js:339]:
```suggestion
  it('should not return a warning when a anyOf model with a nested allOf defines a required property in at least one of its children', () => {
```

File: test/plugins/validation/2and3/items-required-for-array-objects.test.js [hunk 12]
```diff
@@ -190,17 +191,293 @@ describe('validation plugin - semantic - items required for array objects - Open
...
+        }
+      }
+    };
+
+    const res = validate({ jsSpec: spec }, config);
+    expect(res.errors.length).toEqual(0);
+    expect(res.warnings.length).toEqual(0);
+  });
+
+  it('should not return a warning when a model with a nested allOf defines a required property in atleast one of its children', () => {
```
Comment (dpopp07) [inline test/plugins/validation/2and3/items-required-for-array-objects.test.js:389]:
```suggestion
  it('should not return a warning when a model with a nested allOf defines a required property in at least one of its children', () => {
```

File: test/plugins/validation/2and3/items-required-for-array-objects.test.js [hunk 13]
```diff
@@ -190,17 +191,293 @@ describe('validation plugin - semantic - items required for array objects - Open
...
+        }
+      }
+    };
+
+    const res = validate({ jsSpec: spec }, config);
+    expect(res.errors.length).toEqual(0);
     expect(res.warnings.length).toEqual(0);
   });
 
+  it('should return a warning when a model with an allOf does not define a required property in atleast one of its children', () => {
```
Comment (dpopp07) [inline test/plugins/validation/2and3/items-required-for-array-objects.test.js:439]:
```suggestion
  it('should return a warning when a model with an allOf does not define a required property in at least one of its children', () => {
```

File: README.md [hunk 14]
```diff
@@ -213,6 +213,7 @@ The supported rules are described below:
 | no_array_responses           | Flag any operations with a top-level array response.                                | shared   |
 | parameter_order              | Flag any operations with optional parameters before a required param.               | shared   |
 | operation_id_naming_convention | Flag any `operationId` that does not follow naming convention.                    | shared   |
+| undefined_required_properties| Flag any operations with undefined required properties                              | shared   |
```
Comment (dpopp07) [inline README.md:216]: We aren't quite flagging "operations" so I think this description could be more accurate - we're flagging properties of a schema.

This comment is making me realize this configuration option belongs under the "schemas" category rather than the "operations" category - my apologies for not catching this earlier

File: src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js [hunk 15]
```diff
@@ -10,8 +10,47 @@
...
+    return reqPropDefined;
+  } else if (Array.isArray(obj.allOf)) {
+    let reqPropDefined = false;
+    obj.allOf.forEach((childObj) => {
+      if (childObj.properties && childObj.properties[requiredProp]) {
+        reqPropDefined = true;
+      } else if (checkReqProp(jsSpec, childObj, requiredProp)) {
+        reqPropDefined = true;
+      }
+    });
```
Comment (dpopp07) [inline src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js:47]: Just like above, I think this can be simplified:

```suggestion
    obj.allOf.forEach((childObj) => {
      if (checkReqProp(jsSpec, childObj, requiredProp)) {
        reqPropDefined = true;
      }
    });
```

File: test/plugins/validation/2and3/items-required-for-array-objects.test.js (hunk 11 above)
Comment (rmkeezer) [reply test/plugins/validation/2and3/items-required-for-array-objects.test.js:339]: Good catch, I fixed this and must have accidentally reverted it haha
